### PR TITLE
Record-structs: Address some PROTOTYPE markers

### PR DESCRIPTION
**Pull request by jcouv**
Test plan https://github.com/dotnet/roslyn/issues/51199

Fixes https://github.com/dotnet/roslyn/issues/52233 (public API)

### CODE REVIEW

File: src/Compilers/CSharp/CSharpAnalyzerDriver/CSharpDeclarationComputer.cs [hunk 1]
```diff
@@ -117,6 +115,20 @@ private static bool InvalidLevel(int? level)
                             return;
                         }
 
+                        goto case SyntaxKind.ClassDeclaration;
+                    }
+                case SyntaxKind.RecordStructDeclaration:
+                    {
+                        if (associatedSymbol is IMethodSymbol ctor)
+                        {
```
Comment (jcouv) [inline src/Compilers/CSharp/CSharpAnalyzerDriver/CSharpDeclarationComputer.cs:123]: :memo: I didn't manage to cover this `if` and couldn't find coverage for above `if` either.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs [hunk 2]
```diff
@@ -2253,7 +2253,6 @@ ITypeSymbol ITypeSymbolInternal.GetITypeSymbol()
             return GetITypeSymbol(DefaultNullableAnnotation);
         }
 
-        // PROTOTYPE(record-structs): rename to IsRecordClass?
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbol.cs:2256]: :memo: not planning to rename the internal API

File: src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs [hunk 3]
```diff
@@ -795,7 +795,6 @@ public static SyntaxKind GetTypeDeclarationKind(SyntaxKind kind)
                 case SyntaxKind.InterfaceKeyword:
                     return SyntaxKind.InterfaceDeclaration;
                 case SyntaxKind.RecordKeyword:
-                    // PROTOTYPE(record-structs): anything we can do?
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxKindFacts.cs:798]: :memo: I don't see anything we could do given the shape of the API. We'll see if that causes problems during IDE work

File: src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs [hunk 4]
```diff
@@ -2983,7 +2983,6 @@ void verifyParsedAsRecord()
         [Fact, CompilerTrait(CompilerFeature.RecordStructs)]
         public void RecordInterfaceParsing()
         {
-            // PROTOTYPE(record-structs): should we treat this as a binding error instead, for better error recovery?
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RecordParsing.cs:2986]: :memo: Added a bullet to test plan

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs [hunk 5]
```diff
@@ -227,106 +202,38 @@ static void Main()
...
-                // (6,16): error CS0573: 'S': cannot have instance property or field initializers in structs
-                //     public int Z = 0;
-                Diagnostic(ErrorCode.ERR_FieldInitializerInStruct, "Z").WithArguments("S").WithLocation(6, 16)
-                );
         }
 
         [Fact]
         public void RecordStructLanguageVersion()
         {
-            // PROTOTYPE(record-structs): can we improve the error recovery, maybe treating this as a record struct with missing `record`?
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:329]: :memo: Added a bullet to test plan

File: src/Compilers/CSharp/CSharpAnalyzerDriver/CSharpDeclarationComputer.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/CSharpAnalyzerDriver/CSharpDeclarationComputer.cs:123]: > I didn't manage to cover this if and couldn't find coverage for above if either.

Really? I acn easily hit code paths in the previous case by running existing test in src\Compilers\CSharp\Test\Semantic\Semantics\RecordTests.cs

---
In reply to: [604290986](https://github.com/dotnet/roslyn/pull/52256#discussion_r604290986) [](ancestors = 604290986)

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs [hunk 6]
```diff
@@ -1094,17 +1091,10 @@ private MemberSemanticModel CreateMemberModel(CSharpSyntaxNode node)
...
                     {
                         SynthesizedRecordConstructor symbol = TryGetSynthesizedRecordConstructor((RecordDeclarationSyntax)node);
-
-                        if (symbol is null)
-                        {
-                            return null;
-                        }
-
-                        return createMethodBodySemanticModel(node, symbol);
+                        return symbol is null ? null : createMethodBodySemanticModel(node, symbol);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:1097]: >return symbol is null ? null : createMethodBodySemanticModel(node, symbol); [](start = 24, length = 75)

Please revert style change.

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs [hunk 7]
```diff
@@ -2427,6 +2420,7 @@ internal override Symbol RemapSymbolIfNecessaryCore(Symbol symbol)
                         case SymbolKind.NamedType:
                             Debug.Assert((object)declaredSymbol.GetSymbol() == (object)ctor.ContainingSymbol);
                             // Accept nodes that do not match a 'parameter list'/'base arguments list'.
+                            //return (node) => true;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:2423]: >//return (node) => true;  [](start = 28, length = 25)

We do not commit commented out code without a PROTOTYPE comment or an issue tracking the follow-up

File: src/Compilers/CSharp/CSharpAnalyzerDriver/CSharpDeclarationComputer.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/CSharpAnalyzerDriver/CSharpDeclarationComputer.cs:123]: Thanks. You're right. Both `if` blocks are visited by the `AnalyzerActions_XX` tests in `RecordTests` and `RecordStructTests`.

---
In reply to: [605097075](https://github.com/dotnet/roslyn/pull/52256#discussion_r605097075) [](ancestors = 605097075,604290986)

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs [hunk 8]
```diff
@@ -565,7 +565,8 @@ internal override bool IsRecord
             }
         }
 
-        // PROTOTYPE(record-structs): update for record structs (is there a way to recognize a record struct from PE?)
+        // Is there a way to recognize a record struct from PE?
+        // Tracked by https://github.com/dotnet/roslyn/issues/52233
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs:569]: >// Tracked by https://github.com/dotnet/roslyn/issues/52233 [](start = 8, length = 59)

Have we resolved this design question?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 9]
```diff
@@ -515,6 +516,32 @@ record R(R X)
             Assert.Equal("R..ctor(R X)", initializer.ContainingSymbol.ToTestDisplayString());
         }
 
+        [Fact]
+        public void GetDeclaredSymbolOnFieldInitializer()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:520]: >Field [](start = 39, length = 5)

Field or Property?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:520]: And the test doesn't do either.

---
In reply to: [605126785](https://github.com/dotnet/roslyn/pull/52256#discussion_r605126785) [](ancestors = 605126785)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs [hunk 10]
```diff
@@ -5190,5 +5156,959 @@ public void ToString_UserDefinedPrintMembers_Static()
                 Diagnostic(ErrorCode.ERR_StaticAPIInRecord, "PrintMembers").WithArguments("C1.PrintMembers(System.Text.StringBuilder)").WithLocation(4, 25)
                 );
         }
+
+        [Fact]
+        public void AmbigCtor_WithFieldInitializer()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:5161]: >AmbigCtor_WithFieldInitializer [](start = 20, length = 30)

It is not clear why the test is called this. It doesn't look like the scenario has any ambiguity.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs [hunk 11]
```diff
@@ -5190,5 +5156,959 @@ public void ToString_UserDefinedPrintMembers_Static()
...
+
+            var initializerSyntax = tree.GetRoot().DescendantNodes().OfType<EqualsValueClauseSyntax>().Single();
+            var initializer = model.GetSymbolInfo(initializerSyntax.Value).Symbol!;
+            Assert.Equal("R X", initializer.ToTestDisplayString());
+            Assert.Equal(SymbolKind.Parameter, initializer.Kind);
+            Assert.Equal("R..ctor(R X)", initializer.ContainingSymbol.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void GetDeclaredSymbolOnFieldInitializer()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:5192]: >GetDeclaredSymbolOnFieldInitializer [](start = 20, length = 35)

Similar comment as for the test in RecordTest.cs

File: src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs [hunk 12]
```diff
@@ -1094,11 +1091,9 @@ private MemberSemanticModel CreateMemberModel(CSharpSyntaxNode node)
                         return createMethodBodySemanticModel(memberDecl, symbol);
                     }
 
-                // PROTOTYPE(record-structs): update for record structs
                 case SyntaxKind.RecordDeclaration:
                     {
                         SynthesizedRecordConstructor symbol = TryGetSynthesizedRecordConstructor((RecordDeclarationSyntax)node);
-
                         if (symbol is null)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs:1097]: >if (symbol is null) [](start = 24, length = 19)

An empty line was unnecessarily removed above this line.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 2)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs (hunk 10 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:5161]: This is a ported test from record classes where it is an ambiguity. I kept the matching test names when moving them.

---
In reply to: [605132581](https://github.com/dotnet/roslyn/pull/52256#discussion_r605132581) [](ancestors = 605132581)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs (hunk 9 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:520]: Changed test name to "Property".
This test is calling getting a declared symbol (out var) that's part of the intializer.

---
In reply to: [605127667](https://github.com/dotnet/roslyn/pull/52256#discussion_r605127667) [](ancestors = 605127667,605126785)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs [hunk 13]
```diff
@@ -5190,5 +5314,1050 @@ public void ToString_UserDefinedPrintMembers_Static()
...
+                        break;
+                    default:
+                        Assert.True(false);
+                        break;
+                }
+
+                context.RegisterSyntaxNodeAction(Handle1, SyntaxKind.NumericLiteralExpression);
+                context.RegisterSyntaxNodeAction(Handle2, SyntaxKind.EqualsValueClause);
+                context.RegisterSyntaxNodeAction(Fail, SyntaxKind.BaseConstructorInitializer);
+                context.RegisterSyntaxNodeAction(Fail, SyntaxKind.ConstructorDeclaration);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:6187]: >SyntaxKind.ConstructorDeclaration [](start = 55, length = 33)

Why not test these two?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs [hunk 14]
```diff
@@ -5190,5 +5314,1050 @@ public void ToString_UserDefinedPrintMembers_Static()
...
+            => ImmutableArray.Create(Descriptor);
+
+            public override void Initialize(AnalysisContext context)
+            {
+                context.RegisterOperationAction(HandleConstructorBody, OperationKind.ConstructorBody);
+                context.RegisterOperationAction(HandleInvocation, OperationKind.Invocation);
+                context.RegisterOperationAction(HandleLiteral, OperationKind.Literal);
+                context.RegisterOperationAction(HandleParameterInitializer, OperationKind.ParameterInitializer);
+                context.RegisterOperationAction(Fail, OperationKind.PropertyInitializer);
+                context.RegisterOperationAction(Fail, OperationKind.FieldInitializer);
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:5895]: They were not tested for record class (Handle5 there always fails).

---
In reply to: [605217978](https://github.com/dotnet/roslyn/pull/52256#discussion_r605217978) [](ancestors = 605217978)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 6), except some possible reduction in test coverage around analyzer actions by comparison to record classes

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs (hunk 13 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs]: Added. Thanks

---
In reply to: [605218435](https://github.com/dotnet/roslyn/pull/52256#discussion_r605218435) [](ancestors = 605218435)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs [hunk 15]
```diff
@@ -516,19 +425,175 @@ public void TypeDeclaration_IsStruct()
...
+            {
+                var isSourceSymbol = module is SourceModuleSymbol;
+
+                var point = module.GlobalNamespace.GetTypeMember("Point");
+                Assert.True(point.IsValueType);
+                Assert.False(point.IsReferenceType);
+                Assert.False(point.IsRecord);
+                Assert.Equal(TypeKind.Struct, point.TypeKind);
+                Assert.Equal(SpecialType.System_ValueType, point.BaseTypeNoUseSiteDiagnostics.SpecialType);
+                Assert.True(SyntaxFacts.IsTypeDeclaration(SyntaxKind.RecordStructDeclaration));
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:443]: I'm confused by this assert. It seems independent of any particular compilation, but is rather just testing the SyntaxKind constant. Was that intentional?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs [hunk 16]
```diff
@@ -516,19 +425,175 @@ public void TypeDeclaration_IsStruct()
...
+            Assert.IsType<RetargetingNamedTypeSymbol>(point);
             Assert.True(point.IsRecordStruct);
-            Assert.Equal(TypeKind.Struct, point.TypeKind);
-            Assert.Equal(SpecialType.System_ValueType, point.BaseTypeNoUseSiteDiagnostics.SpecialType);
+            Assert.True(point.GetPublicSymbol().IsRecord);
+        }
 
-            Assert.True(SyntaxFacts.IsTypeDeclaration(SyntaxKind.RecordStructDeclaration));
+        [Fact]
+        public void IsRecord_AnonymousType()
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:515]: Do these and the following new tests belong in the RecordTests.cs, not in RecordStructTests.cs?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs [hunk 17]
```diff
@@ -5190,5 +5314,1050 @@ public void ToString_UserDefinedPrintMembers_Static()
...
         }
+
+        [Fact]
+        public void AmbigCtor_WithPropertyInitializer()
+        {
+            // Scenario causes ambiguous ctor for record class, but not record struct
+            var src = @"
+record struct R(R X)
+{
+    public R X { get; init; } = X;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:5325]: Is there also a test for when no property is explicitly declared?

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs (hunk 15 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:443]: Yes, it's just testing that SyntaxFacts API :-) 
Note that although it appears as a change in the diff, this is an existing check.

---
In reply to: [606013588](https://github.com/dotnet/roslyn/pull/52256#discussion_r606013588) [](ancestors = 606013588)

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:443]: I'll move it out of the `validateModule` local function

---
In reply to: [606392703](https://github.com/dotnet/roslyn/pull/52256#discussion_r606392703) [](ancestors = 606392703,606013588)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs (hunk 16 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:515]: Could do either way.

---
In reply to: [606013947](https://github.com/dotnet/roslyn/pull/52256#discussion_r606013947) [](ancestors = 606013947)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs (hunk 17 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordStructTests.cs:5325]: Added. Thanks

---
In reply to: [606014509](https://github.com/dotnet/roslyn/pull/52256#discussion_r606014509) [](ancestors = 606014509)

**Pull request merged**